File: 77-组合.cpp
class Solution {
public:
    vector<vector<int>> combine(int n, int k) {
        vector<vector<int>> result;
        vector<int> myvector;
        
        dfs(result, myvector, n, k);

        return result;
    }

    void dfs(vector<vector<int>>& result, vector<int>& myvector, int n, int k, int num = 0, int pre = 0){
        if(n - pre < k - num)      // 数量不够，剪枝
            return;
        
        if(num == k){
            result.push_back(myvector);
            return;
        }
        
        for(int i = pre + 1; i < n + 1; i++){
            myvector.push_back(i);
            dfs(result, myvector, n, k, num + 1, i);
            myvector.pop_back();
        }
    }
};

File: 77-组合.py
class Solution:
    def combine(self, n: int, k: int) -> List[List[int]]:
        result = []
        mylist = []
        myset = set()

        def dfs(num=0, pre=0):
            if n - pre < k - num:             ## 数字数量不够，剪枝
                return

            if num == k:
                result.append(mylist[:])
                return
            
            for i in range(pre + 1, n + 1):     ## 为避免重复，序列保持有序，后面的比pre大  
                mylist.append(i)
                dfs(num + 1, i)
                mylist.pop()
        
        dfs()

        return result
